agent/agent.go

```diff
@@ -2135,68 +2124,77 @@ func (a *Agent) AddServiceAndReplaceChecks(service *structs.NodeService, chkType
 func (a *Agent) AddService(service *structs.NodeService, chkTypes []*structs.CheckType, persist bool, token string, source configSource) error {
 	a.stateLock.Lock()
 	defer a.stateLock.Unlock()
-	return a.addServiceLocked(service, chkTypes, nil, true, persist, true, token, false, source)
+	return a.addServiceLocked(&addServiceRequest{
+		service:               service,
+		chkTypes:              chkTypes,
+		previousDefaults:      nil,
+		waitForCentralConfig:  true,
+		persist:               persist,
+		persistServiceConfig:  true,
+		token:                 token,
+		replaceExistingChecks: false,
+		source:                source,
+	})
 }
 
 // addServiceLocked adds a service entry to the service manager if enabled, or directly
 // to the local state if it is not. This function assumes the state lock is already held.
-func (a *Agent) addServiceLocked(
-	service *structs.NodeService,
-	chkTypes []*structs.CheckType,
-	previousDefaults *structs.ServiceConfigResponse,
-	waitForCentralConfig bool,
-	persist bool,
-	persistServiceConfig bool,
-	token string,
-	replaceExistingChecks bool,
-	source configSource,
-) error {
-	if err := a.validateService(service, chkTypes); err != nil {
+func (a *Agent) addServiceLocked(req *addServiceRequest) error {
+	req.fixupForAddServiceLocked()
+
+	if err := a.validateService(req.service, req.chkTypes); err != nil {
 		return err
 	}
 
 	if a.config.EnableCentralServiceConfig {
-		return a.serviceManager.AddService(
-			service,
-			chkTypes,
-			previousDefaults,
-			waitForCentralConfig,
-			persist,
-			persistServiceConfig,
-			token,
-			replaceExistingChecks,
-			source,
-		)
+		return a.serviceManager.AddService(req)
 	}
 
 	// previousDefaults are ignored here because they are only relevant for central config.
-	return a.addServiceInternal(&addServiceRequest{
-		service:               service,
-		chkTypes:              chkTypes,
-		persistService:        nil,
-		persistDefaults:       nil,
-		persist:               persist,
-		persistServiceConfig:  false,
-		token:                 token,
-		replaceExistingChecks: replaceExistingChecks,
-		source:                source,
-	})
+	req.persistService = nil
+	req.persistDefaults = nil
+	req.persistServiceConfig = false
+
+	return a.addServiceInternal(req)
 }
 
+// addServiceRequest is the union of arguments for calling both
+// addServiceLocked and addServiceInternal. The overlap was significant enough
+// to warrant merging them and indicating which fields are meant to be set only
+// in one of the two contexts.
+//
+// Before using the request struct one of the fixupFor*() methods should be
+// invoked to clear irrelevant fields.
+//
+// The ServiceManager.AddService signature is largely just a passthrough for
+// addServiceLocked and should be treated as such.
```

Comment:
🤔 my initial reaction to this was that is seems uglier than the massive argument lists it replaced and maybe the smell indicated that we should have just built this whole registration pipeline differently.

But after thinking some I'm not sure it's too bad and the alternatives don't seem obviously cleaner. I think it's OK as it is.

Reply:
If we start really using central service config more it might be worth the investment to rethink this whole `Agent` structure. It's not a quick fix.